World Hopper plugin - Type filter

Blackberry0Pie opened the pull request:
Because why not.
example config:
![config](https://i.gyazo.com/4233486b251ed131099ddaa3b7a532d0.png)
example plugin panel
![league_example](https://i.gyazo.com/520b2dfb2e380aef18f7ecf223f262d1.png)

dekvall commented:
Related to #435

Adam- commented:
1) I think it would be better if this could show potentially more than one type instead of just one type. Maybe just make an option per type
2) I see that this doesn't affect the world hopper next/prev feature, which it probably should

Blackberry0Pie commented:
Updated to have a filter for each type.
Added a toggle for filter exclusivity(not showing world types that are not selected).
Changed behavior of world hopper regarding {skill total, bounty, and LMS} worlds when exclusivity is true.

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/worldhopper/WorldHopperConfig.java:
```diff
@@ -125,11 +125,110 @@ default SubscriptionFilterMode subscriptionFilter()
+	default boolean PVPTypeFilter()
+{
+	return false;
+}
```
These lines should be indented one level further.

Blackberry0Pie commented:
added config section as per request

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/worldhopper/WorldSwitcherPanel.java:
```diff
@@ -247,6 +266,67 @@ void populate(List<World> worlds)
 
+			if (exclusiveTypeFilter)
+			{
+				if (!world.getTypes().contains(WorldType.PVP) &&
```
Our code convention for multiline boolean operations are to put the operators at the start of continued lines like so:

```java
a
	&& b
	&& c;
```

Nightfirecat replied:
Additionally, this check for `world.getTypes()` containing at least one of the many types here can be implemented more cleanly using [`Collections.disjoint(A, B)`](https://docs.oracle.com/javase/8/docs/api/java/util/Collections.html#disjoint-java.util.Collection-java.util.Collection-)

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/worldhopper/WorldHopperConfig.java:
```diff
@@ -129,18 +139,127 @@ default boolean menuOption()
+	@ConfigItem(
+		keyName = "exclusiveTypeFilter",
+		name = "Exclusive filtering",
+		description = "Only show selected world types",
```
This is not clear enough to be understandable. Without looking at the code implemented using this toggle, I cannot understand what behavior this changes.

Blackberry0Pie replied:
Requested changes done. Hopefully this description is now understandable.

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/worldhopper/WorldHopperConfig.java:
```diff
@@ -30,11 +31,20 @@
+	@ConfigSection(
+		name = "World Type Filtering",
+		description = "Filtering options for world types",
+		position = -1,
```
Maybe this should appear at the end of the config list?

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/worldhopper/WorldSwitcherPanel.java:
```diff
@@ -230,23 +253,88 @@ void populate(List<World> worlds)
+			if (exclusiveTypeFilter)
+			{
+				//this removes worlds that have no type as they will not match any filter(s)
+				EnumSet<WorldType> wt = EnumSet.allOf(WorldType.class);
```
No need for the shorthand, give this variable a clearer name. Maybe `allWorldTypes`?

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/worldhopper/WorldSwitcherPanel.java:
```diff
@@ -230,23 +253,88 @@ void populate(List<World> worlds)
 
+			if (exclusiveTypeFilter)
+			{
+				//this removes worlds that have no type as they will not match any filter(s)
```
This should be above the `Collections.disjoint()` line, not the line that defines the set of all world types.

Nightfirecat commented:
This still doesn't seem to affect the prev/next world hop keybind. This was tested with exclusive filtering on, and only "Show Skill Total" selected under the filtering section.

Nightfirecat commented:
Now that there's been a pretty significant other change (the last commit looks purely like a separate bugfix?) it would be helpful to squash and rebase this history into atomic commits for these changes--one for the panel bugfix, and the second for the type filter.

Adam- commented:
added in db7ae8d62d57f421ba72566be8bd777a8a76452a
